docs/dev/json.rst

```diff
@@ -177,7 +177,10 @@ element                 description
 ``isrc``                The ISRC code associated with the recording.
 ``spotify_id``          The Spotify track URL associated with this recording.  e.g.: http://open.spotify.com/track/1rrgWMXGCGHru5bIRxGFV0
 ``tags``                A list of user defined tags to be associated with this recording. These tags are similar to last.fm tags. For example, you have apply tags such as ``punk``, ``see-live``, ``smelly``. You may submit up to :data:`~webserver.views.api.MAX_TAGS_PER_LISTEN` tags and each tag may be up to :data:`~webserver.views.api.MAX_TAG_SIZE` characters large.
-``listening_from``       The source of the listen, i.e the name of the client or service which submits the listen.
+``listening_from``      The name of the client which submits the listen.
+``listening_from_version`` (optional). If the client submitting the listen has a version, put it here.
+``source``              If the song of this listen comes from an online source, the name of this source (e.g. Spotify, Soundcloud, YouTube, Jamendo). This should be a name, not a URL.
+``origin_url``          If the song of this listen comes from an online source, the URL to the place where it is available. This could be a spotify url (see ``spotify_id``), a YouTube video URL, a Soundcloud recording page URL, or the full URL to a public MP3 file.
```

Comment:
```suggestion
``origin_url``          If the song of this listen comes from an online source, the URL to the place where it is available. This could be a spotify url (see ``spotify_id``), a YouTube video URL, a Soundcloud recording page URL, or the full URL to a public MP3 file. This must be a valid public URL rather than a resource ID or URI.
```
?

Reply:
so if we're going to be explicit, then perhaps we should add more json examples, showing the case of youtube, or soundcloud, or jamendo, or archive.org

Reply:
Here's an abridged example of what I'm going to do for BrainzPlayer in case it's useful:
```json
"track_metadata": {
  "additional_info": {
    "listening_from": "listenbrainz",
    "origin_url": "https://www.youtube.com/watch?v=Bq7-B8_NZNU",
    "source": "youtube"
  }
}
```

Reply:
What about having the fields named `source` and `source_url` or `origin` and `origin_url`? As I see it they are related, and source and origin kind of mean the same here.

Reply:
Can you explain what each of these 4 fields would mean (perhaps with some examples?)

Reply:
I think your description in this PR already fits this very well, we only need to get the naming of the fields right. And I agree that it's better to get this right now then to wait. Taking into account Rob's suggestion of using `listening_client` and using a consistent naming for origin and origin URL I would suggest something like this:

```
"track_metadata": {
  "additional_info": {
    "listening_client": "Web Scrobbler",
    "origin": "bandcamp",
    "origin_url": "https://paradiselostofficial.bandcamp.com/track/the-longest-winter"
  }
}
```

I am also not sure about the need of `listening_client_version`, but if someone wants it I see no harm of having it.